pkg/broker/cel.go

```diff
@@ -0,0 +1,137 @@
+package broker
+
+import (
+	"bytes"
+	"encoding/json"
+
+	"github.com/cloudevents/sdk-go/pkg/cloudevents"
+	"github.com/golang/protobuf/jsonpb"
+	structpb "github.com/golang/protobuf/ptypes/struct"
+	"github.com/google/cel-go/cel"
+	"github.com/google/cel-go/checker/decls"
+	"github.com/google/cel-go/common/types"
+	eventingv1alpha1 "github.com/knative/eventing/pkg/apis/eventing/v1alpha1"
+	celprotos "github.com/knative/eventing/pkg/broker/dev_knative"
+	"go.uber.org/zap"
+)
+
+const (
+	// CELVarKeyContext is the CEL variable key used for the CloudEvent event
+	// context fields defined in the spec.
+	CELVarKeyContext = "ce"
+	// CELVarKeyExtensions is the CEL variable key used for the CloudEvent event
+	// context extensions.
+	CELVarKeyExtensions = "ext"
+	// CELVarKeyData is the CEL variable key used for the CloudEvent event data.
+	CELVarKeyData = "data"
+	//TODO add a key that contains both the extensions and the baseline context
+	// vars so extensions can be future proofed.
+)
+
+func (r *Receiver) filterEventByCEL(ts *eventingv1alpha1.TriggerSpec, event *cloudevents.Event) (bool, error) {
+	e, err := cel.NewEnv(
+		cel.Types(&celprotos.CloudEventContext{}),
+		cel.Declarations(
+			decls.NewIdent(CELVarKeyContext, decls.NewObjectType("dev.knative.CloudEventContext"), nil),
+			decls.NewIdent(CELVarKeyExtensions, decls.NewObjectType("google.protobuf.Struct"), nil),
+			decls.NewIdent(CELVarKeyData, decls.NewObjectType("google.protobuf.Struct"), nil),
+		),
+	)
+	if err != nil {
+		return false, err
+	}
+
+	p, iss := e.Parse(ts.Filter.CEL.Expression)
+	if iss != nil && iss.Err() != nil {
+		return false, iss.Err()
+	}
+	c, iss := e.Check(p)
+	if iss != nil && iss.Err() != nil {
+		return false, iss.Err()
+	}
+
+	//TODO cache these by hash of expression. Programs are thread-safe so it's
+	// ok to share them between triggers and events.
+	prg, err := e.Program(c)
+	if err != nil {
+		return false, err
+	}
+
+	vars := map[string]interface{}{}
+	// Set baseline context fields
+	ceCtx := &celprotos.CloudEventContext{
+		Specversion:     event.Context.GetSpecVersion(),
+		Type:            event.Context.GetType(),
+		Source:          event.Context.GetSource(),
+		Schemaurl:       event.Context.GetSchemaURL(),
+		Datamediatype:   event.Context.GetDataMediaType(),
+		Datacontenttype: event.Context.GetDataContentType(),
+	}
+	vars[CELVarKeyContext] = ceCtx
+
+	// If the Trigger has requested parsing of extensions, attempt to turn them
+	// into a dynamic struct.
+	if ts.Filter.CEL.ParseExtensions {
+		//TODO should this coerce to V02?
+		extStruct, err := ceParsedExtensionsStruct(event.Context.AsV02().Extensions)
+		if err != nil {
+			r.logger.Error("Failed to parse event context for CEL filtering", zap.String("id", event.Context.AsV02().ID), zap.Error(err))
```

Comment:
Shouldn't this return the err?

Reply:
There are two options when a parse error occurs:
1. Ignore the error and evaluate the expression anyway with the data we parsed correctly.
1. Return the error without evaluating. The filter will reject any event that fails to parse.

The code currently implements 1, assuming that if evaluation is possible, it should be attempted. Thus the error is not returned here, because that would skip the expression evaluation. It would probably be a good idea to save the error and return it along with the expression evaluation result.

Here's an example of a Trigger that probably wants to evaluate anyway even if extensions cannot be parsed:

```yaml
spec:
  filter:
    cel:
      expression: ce.type == "com.example.event"
      parseExtensions: true
```

A failure parsing extensions might not be what the user expects, since the expression could evaluate correctly regardless of whether extensions are successfully parsed.

If we implement option 2, I'd probably rename the field `MustParseExtensions` or something to indicate that a parse failure causes a rejection of the event.

@n3wscott suggested earlier that it may be possible to detect automatically whether the expression requires parsing of extensions or data by inspecting the CEL AST. This seems like the best case for users, but some users may still want to turn parsing behavior on or off independent of the expression.

One possibly terrible option is to make `parseExtensions` a string field, with initial values `yes` and `no` (the default). Future values could be added: `auto` (when AST inspection is possible, this becomes the default), `required` (to indicate behavior on error).